designs/2024-deprecated-rule-metadata/README.md

```diff
@@ -0,0 +1,250 @@
+
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+
+There are long-time rule [properties](https://eslint.org/docs/latest/extend/custom-rules#rule-structure) `meta.deprecated` and `meta.replacedBy` that have been intended to document when rules are deprecated and what their replacement rule(s) are. For the most part, usage would look something like this:
+
+```js
+module.exports = { meta: { deprecated: true, replacedBy: ['replacement-rule-name'] } };
+```
+
+These properties are often used for generating plugin/rule documentation websites and in documentation tooling like [eslint-doc-generator](https://github.com/bmish/eslint-doc-generator).
+
+But there are some limitations to this current format:
+
+- Simply providing the replacement rule name as a string doesn't yield much context/explanation of the replacement/deprecation. That means documentation tooling / websites and code editors can only generate limited information to present about the situation.
+- Some rules provide the replacement rule name with the plugin prefix as in `prefix/rule-name` while others just provide it as `rule-name`, which can result in ambiguity about whether the replacement rule is in the same plugin, a different third-party plugin, or ESLint core. And for third-party plugins, there's no easy way to automatically determine where their documentation is located or how to link to them.
+
+## Detailed Design
+
+<!--
+   This is the bulk of the RFC.
+
+   Explain the design with enough detail that someone familiar with ESLint
+   can implement it by reading this document. Please get into specifics
+   of your approach, corner cases, and examples of how the change will be
+   used. Be sure to define any new terms in this section.
+-->
+
+We propose extended `meta.deprecated` and `meta.replacedBy` rule property schemas to reduce ambiguity and allow additional key details to be represented, described below as a TypeScript type for clarity:
+
+```ts
+type RuleMeta = {
+  deprecated:
+    | boolean // Existing boolean option, backwards compatible.
+    | string // Shorthand property for general deprecation message, such as why the deprecation occurred. Any truthy value implies deprecated.
+    | {
+        message?: string; // General deprecation message, such as why the deprecation occurred.
+        url?: string; // URL to more information about this deprecation in general.
+      };
+  replacedBy:
+    | readonly string[] // Existing shorthand property for rule names, backwards compatible. It's recommended to omit the plugin prefix from rule names.
+    | readonly {
+        /**
+         * Plugin containing the replacement.
+         * Use "eslint" if the replacement is an ESLint core rule.
+         * Omit if the replacement is in the same plugin.
+         */
+        plugin?:
+          | string // Shorthand property for the plugin name i.e. "eslint-plugin-example" that contains the replacement rule.
```

Comment:
[Clarification] Are both `"eslint-plugin-example"` and `"example"` allowed?

Reply:
By "shorthand property", I just mean that a string can be provided instead of the more detailed object format.

My intention was for the full plugin name to be provided here `eslint-plugin-foo` or `@foo/eslint-plugin` or whatever so there wouldn't be any ambiguity. Simply providing the plugin prefix `foo` doesn't necessarily tell us the full plugin name as there can be various formats...